File: cpp/Binary Tree/L_diameterOfBT.cpp
#include<bits/stdc++.h>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

int diameterOfBinaryTree(TreeNode* root) {
    int res = 0;
    dfs(root, res);
    return res;
}

int dfs(TreeNode* root, int& res) {
    if(!root) return 0;
    int left = dfs(root->left, res);
    int right = dfs(root->right, res);
    res = max(res, left + right);
    return 1 + max(left, right);
}
int main() {

}

File: Python/Binary Trees/L_diameterOfBT.py
def diameterOfBinaryTree(root) -> int:
        res = 0
        def dfs(root):
            if not root: return 0
            left = dfs(root.left)
            right = dfs(root.right)
            nonlocal res
            res = max(res, left + right)
            return 1 + max(left, right)
        dfs(root)
        return res
    
def diameterOfBinaryTree1(root) -> int:
        diameter = [0]
        def helper(node):
            if not node:
                return 0
            left = helper(node.left)
            right = helper(node.right)
            diameter[0] = max(diameter[0], left + right)
            return max(left, right) + 1
        helper(root)
        return diameter[0]